### PR TITLE
Update extractor for new 'Short description' format

### PR DESCRIPTION
**Pull request by terhorstd**
I updated the doc extractor as discussed, to rewrite the short descriptions to proper headers for the files. The generated keyword indices now are also interlinked and have shown many minor inconsistencies in user documentation that are fixed. The actual reworking of the documentation (keyword) structure should however be done in a separate PR.

### CODE REVIEW

File: nestkernel/recording_backend_ascii.h [hunk 1]
```diff
@@ -32,8 +32,13 @@
...
 .. _ascii_backend:
 
+Short description
++++++++++++++++++
+
 Write data to plain text files
-##############################
+
+Description
++++++++++++
```
Comment (jougs) [inline nestkernel/recording_backend_ascii.h]: Please undo the changes to this file. `RecordingBackend`s are not models. They still contain user-level documentation that needs to be extracted by the doc extractor, but without the section mangling applied to models. The files are made visible by inclusion into the [guide to recording from simulations](https://github.com/nest/nest-simulator/blob/master/doc/guides/recording_from_simulations.rst).

File: nestkernel/recording_backend_memory.h [hunk 2]
```diff
@@ -28,8 +28,13 @@
...
 /* BeginUserDocs: recording backend
 
+Short description
++++++++++++++++++
+
 Store data in main memory
-#########################
+
+Description
++++++++++++
```
Comment (jougs) [inline nestkernel/recording_backend_memory.h]: Please undo. See above.

File: nestkernel/recording_backend_screen.h [hunk 3]
```diff
@@ -28,8 +28,13 @@
...
 /* BeginUserDocs: recording backend
 
+Short description
++++++++++++++++++
+
 Write data to the terminal
-##########################
+
+Description
++++++++++++
```
Comment (jougs) [inline nestkernel/recording_backend_screen.h]: Please undo. See above.

File: nestkernel/recording_backend_sionlib.h [hunk 4]
```diff
@@ -33,8 +33,13 @@
...
 .. _sionlib_backend:
 
+Short description
++++++++++++++++++
+
 Store data to an efficient binary format
-########################################
+
+Description
++++++++++++
```
Comment (jougs) [inline nestkernel/recording_backend_sionlib.h]: Please undo. See above.

File: models/weight_recorder.h [hunk 5]
```diff
@@ -34,10 +34,12 @@
 #include "nest_types.h"
 #include "recording_device.h"
 
-/* BeginUserDocs: device, recorder
+/* BeginUserDocs: device, recorder, synapse
```
Comment (jougs) [inline models/weight_recorder.h]: I'm unsure about this change, as this is not a synapse model and should thus technically not be tagged synapse in the same way that none of the devices for recording/stimulating neurons is tagged `neuron`.

File: models/spike_detector.h [hunk 6]
```diff
@@ -33,10 +33,12 @@
 #include "nest_types.h"
 #include "recording_device.h"
 
-/* BeginUserDocs: device, recorder, spike
+/* BeginUserDocs: device, recorder, spike, detector
```
Comment (jougs) [inline models/spike_detector.h]:
```suggestion
/* BeginUserDocs: device, recorder, spike
```

**Review by jougs — CHANGES_REQUESTED** [review]
Many thanks! This is really a great addition. I'm much looking forward to when this goes live. I have some minor comments on the changes to documentation, but am totally happy with the code.

File: models/weight_recorder.h (hunk 5 above)
Comment (terhorstd) [reply models/weight_recorder.h]: As discussed, I reverted various doc content changes, as this will be fixed by other PRs and does not concern the extractor per se, so this would also not go into this PR.

**Comment by terhorstd** [comment]
@jougs: you can have another go at this. somehow the re-request button doesn't work for me, maybe only @sarakonradi can do that.

**Comment by sarakonradi** [comment]
@terhorstd @jougs The button appears for me, but it doesn't respond when I click it. Not sure why this is.

File: doc/extractor_userdocs.py [hunk 7]
```diff
@@ -94,32 +97,140 @@ def UserDocExtractor(
...
+    str
+        original parameter doc with see_also section replaced
+    '''
+
+    titles = getTitles(doc)
+    if not titles:
+        raise ValueError("No sections found in '%s'!" % filename)
+
+    def rightcase(text):
+        ' make text title-case except for acronyms '
```
Comment (jougs) [inline doc/extractor_userdocs.py]: please make this a proper docstring enclosed in triple-quotes.

File: doc/extractor_userdocs.py [hunk 8]
```diff
@@ -94,32 +97,140 @@ def UserDocExtractor(
...
+            + fixed_title + "\n" + "=" * len(fixed_title) + "\n\n"
+            + doc[secend:]
+            )
+    raise ValueError("No section '%s' found in %s!", short_description, filename)
+
+
+def rewrite_see_also(doc, filename, tags, see_also="See also"):
+    '''
+    Modify a given text by replacing the first section named as given in
+    `see_also` by the filename and content of that section.
```
Comment (jougs) [inline doc/extractor_userdocs.py]:
```suggestion
    Replace the content of a section named `see_also` in the document `doc` by links
    to all tag-indices, the file `filename` is part of itself.
```

Oder so ähnlich...

File: doc/extractor_userdocs.py [hunk 9]
```diff
@@ -94,32 +97,140 @@ def UserDocExtractor(
...
+    doc : str
+      restructured text with all sections
+
+    filename : str, path
+      name that is inserted in the replaced title (and used for useful error
+      messages).
+
+    see_also : str
+      title of the section that is to be rewritten to the document title
+
```
Comment (jougs) [inline doc/extractor_userdocs.py:207]: Parameter `tags` is not explained. Please do so!

File: models/cont_delay_connection.h [hunk 10]
```diff
@@ -33,7 +33,7 @@
 namespace nest
 {
 
-/* BeginUserDocs: synapse, continuous delay
+/* BeginUserDocs: synapse, connection with delay, continuous
```
Comment (jougs) [inline models/cont_delay_connection.h]: Please don't change tags in this PR. This is for another time.
```suggestion
/* BeginUserDocs: synapse, continuous delay
```

File: models/multimeter.h [hunk 11]
```diff
@@ -38,10 +38,12 @@
 #include "dictutils.h"
 #include "name.h"
 
-/* BeginUserDocs: device, recorder
+/* BeginUserDocs: device, recorder, continuous
```
Comment (jougs) [inline models/multimeter.h]:
```suggestion
/* BeginUserDocs: device, recorder
```

Although I know what you want to say, I think `continuous` is not a good tag. IMO, this is more misleading than helpful. Does it mean continuous threshold detection (what we usually call *precise*), spikes on non-grid time points (usually referred to as *off-grid spiking*, measurements of continuous (i.e. non-spike) properties (which is the case here) or, or, or. The term just has too many different meaning to be a good tag.

File: models/music_cont_in_proxy.h [hunk 12]
```diff
@@ -47,7 +47,7 @@
 namespace nest
 {
 
-/* BeginUserDocs: device, MUSIC
+/* BeginUserDocs: device, MUSIC, continuous
```
Comment (jougs) [inline models/music_cont_in_proxy.h]:
```suggestion
/* BeginUserDocs: device, MUSIC
```

Another meaning to the max! This MUSIC proxy receives an array of floating point numbers per time step from somewhere else. The user can read them out of the status dict and do with them whatever she pleases to do. I don't really see a unified feature with the other things tagged `continuous` other than that they all operate on floats somehow.

File: models/music_cont_out_proxy.h [hunk 13]
```diff
@@ -51,7 +51,7 @@
 namespace nest
 {
 
-/* BeginUserDocs: device, MUSIC
+/* BeginUserDocs: device, MUSIC, continuous
```
Comment (jougs) [inline models/music_cont_out_proxy.h]:
```suggestion
/* BeginUserDocs: device, MUSIC
```

File: models/music_event_in_proxy.h [hunk 14]
```diff
@@ -40,7 +40,7 @@
 namespace nest
 {
 
-/* BeginUserDocs: devices, MUSIC
+/* BeginUserDocs: device, MUSIC, spike
```
Comment (jougs) [inline models/music_event_in_proxy.h:43]: :+1:

This is different from *continuous* in the that these proxies actually receive and send spike events from/to neurons and thus have to be listed in some index together with them.

File: models/music_rate_in_proxy.h [hunk 15]
```diff
@@ -42,15 +42,12 @@
 // Includes from sli:
 #include "arraydatum.h"
 
-/* BeginUserDocs:
+/* BeginUserDocs: device, rate, MUSIC
```
Comment (jougs) [inline models/music_rate_in_proxy.h:45]: :+1: because of the same reason as for `music_event_*_proxy`, just with rate events instead of spike events.

**Review by jougs — CHANGES_REQUESTED** [review]
I am so happy that this becomes true after all these years! Yay! Still some comments. Sorry!

File: doc/extractor_userdocs.py (hunk 8 above)
Comment (terhorstd) [reply doc/extractor_userdocs.py]: good catch
